oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java

```diff
@@ -0,0 +1,217 @@
+                 .name(name)
+                 .hardware(hardwareRequest)
+                 .rsaKey(options.getPublicKey())
+                 .applianceId(image.getId())
+                 .dataCenterId(dataCenterId)
+                 .powerOn(Boolean.TRUE).build();
+
+         logger.trace("<< provisioning server '%s'", serverRequest);
+
+         server = api.serverApi().create(serverRequest);
+         logger.trace(">> provisioning complete for server. returned id='%s'", server.id());
+
+      } catch (Exception ex) {
+         logger.error(ex, ">> failed to provision server. rollbacking..");
+         throw Throwables.propagate(ex);
+      }
+      LoginCredentials serverCredentials;
+      
+      //if no private key was set use the set/generated password.
+      if (privateKey == null) {
+         serverCredentials = LoginCredentials.builder()
+                 .user(loginUser)
+                 .password(password)
+                 .build();
+      } else {
+         serverCredentials = LoginCredentials.builder()
+                 .user(loginUser)
+                 .privateKey(privateKey)
+                 .build();
+
+      }
+
+      return new NodeAndInitialCredentials<Server>(server, server.id(), serverCredentials);
+
+   }
+
+   @Override
+   public List<HardwareFlavour> listHardwareProfiles() {
+      return api.serverApi().listHardwareFlavours();
+   }
+
+   @Override
+   public Iterable<Image> listImages() {
+      return api.imageApi().list();
+   }
+
+   @Override
+   public Image getImage(String id) {
+      // try search images
+      logger.trace("<< searching for image with id=%s", id);
+      Image image = api.imageApi().get(id);
+      if (image != null) {
+         logger.trace(">> found image [%s].", image.name());
+         return image;
+      }
+      throw new ResourceNotFoundException("No image with id '" + id + "' was found");
+   }
+
+   @Override
+   public Iterable<Location> listLocations() {
+      // Will never be called
+      throw new UnsupportedOperationException("Locations are configured in jclouds properties");
+   }
+
+   @Override
+   public Server getNode(String id) {
+      return api.serverApi().get(id);
+   }
+
+   @Override
+   public void destroyNode(String id) {
+      waitServerUntilAvailable.apply(getNode(id));
+      logger.debug("Destroying %s ...", id);
+      api.serverApi().delete(id);
+      logger.debug("Destroyed %s!", id);
+   }
+
+   @Override
+   public void rebootNode(String id) {
+      waitServerUntilAvailable.apply(getNode(id));
+      api.serverApi().updateStatus(id, Server.UpdateStatus.create(Types.ServerAction.REBOOT, Types.ServerActionMethod.HARDWARE));
+   }
+
+   @Override
+   public void resumeNode(String id) {
+      api.serverApi().updateStatus(id, Server.UpdateStatus.create(Types.ServerAction.POWER_ON, Types.ServerActionMethod.HARDWARE));
+   }
+
+   @Override
+   public void suspendNode(String id) {
+      waitServerUntilRunning.apply(getNode(id));
```

Comment:
What if we call this on an already suspended node? Will it block until it times out?
Perhaps it is a better idea to validate the state as a precondition and fail, or not validate it at all and let the it fail in the server side.